### PR TITLE
fix(display): fix display number of days

### PR DESCRIPTION
**Pull request by stonebuzz**
```autopurge_delay```  field from ```Entity``` class may exceed a thousand

image

From GLPI entity list, ```autopurge_delay``` field is truncated

image

 if contains space or comma , depending of GLPI configuration

image

Because GLPI use ```sprintf``` function with ```%d```
https://github.com/glpi-project/glpi/blob/7c5796107d7d3c1d8ea02d5b5cb0b0f8d59603f5/inc/dropdown.class.php#L1534

Called from here
https://github.com/glpi-project/glpi/blob/7c5796107d7d3c1d8ea02d5b5cb0b0f8d59603f5/inc/search.class.php#L6437-L6439

This PR proposed to use ```days``` which does not exist  
to use ```default``` case

https://github.com/glpi-project/glpi/blob/7c5796107d7d3c1d8ea02d5b5cb0b0f8d59603f5/inc/dropdown.class.php#L1559-L1560

| Q             | A
| ------------- | ---
| Bug fix?      | yes/no
| New feature?  | yes/no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | internal 22126

### CODE REVIEW

**Review by trasher — APPROVED** [review]
I do not know if there is another way to fix that; ok for me

**Review by orthagh — APPROVED** [review]

**Review by cedric-anne — COMMENTED** [review]
IMHO, we should not introduce a `rawnumber` to handle this case.
Indeed, I think we should:
1. Remove the call to `Html::formatNumber()` in count/number rendering.
2. Add a call to `Html::formatNumber()` in `Dropdown::getValueWithUnit()`.

For example:
```php
   static function getValueWithUnit($value, $unit, $decimals = null) {

      if (strlen($unit) == 0) {
         return $value;
      }

      $formatted_number = Html::formatNumber($data[$ID][$k]['name'], false, $decimals !== null ? $decimals : -1);

      switch ($unit) {
         case 'year' :
            //TRANS: %s is a number of years
            return sprintf(_n('%s year', '%s years', $value), $formatted_number);

...

      }
   }

**Comment by stonebuzz** [comment]
> IMHO, we should not introduce a `rawnumber` to handle this case.
> Indeed, I think we should:
> 
>     1. Remove the call to `Html::formatNumber()` in count/number rendering.
> 
>     2. Add a call to `Html::formatNumber()` in `Dropdown::getValueWithUnit()`.
> 
> 
> For example:
> 
> ```
>    static function getValueWithUnit($value, $unit, $decimals = null) {
> 
>       if (strlen($unit) == 0) {
>          return $value;
>       }
> 
>       $formatted_number = Html::formatNumber($data[$ID][$k]['name'], false, $decimals !== null ? $decimals : -1);
> 
>       switch ($unit) {
>          case 'year' :
>             //TRANS: %s is a number of years
>             return sprintf(_n('%s year', '%s years', $value), $formatted_number);
> 
> ...
> 
>       }
>    }
> ```

args need to be inverted no ?

```
return sprintf(_n('%d day', '%d days', $formatted_number), $value);
```

If not, this method does not work, because ```Html::formatNumber``` add space or comma (depending of GLPI config)
ans ```sprintf``` use ```%d``` arg

**Comment by cedric-anne** [comment]
> If not, this method does not work, because `Html::formatNumber` add space or comma (depending of GLPI config)
> ans `sprintf` use `%d` arg

The numeric value has to be passed to _n to be sure to use most suitable singular/plural form. I changed `%d` to `%s` in the sprintf method, as I think we should use configured number format for all serach results.

**Comment by stonebuzz** [comment]
> > If not, this method does not work, because `Html::formatNumber` add space or comma (depending of GLPI config)
> > ans `sprintf` use `%d` arg
> 
> The numeric value has to be passed to _n to be sure to use most suitable singular/plural form. I changed `%d` to `%s` in the sprintf method, as I think we should use configured number format for all search results.

my mistake, I'm trying

**Comment by stonebuzz** [comment]
OK for me
image

File: inc/dropdown.class.php [hunk 1]
```diff
@@ -1513,41 +1513,44 @@ static function showNumber($myname, $options = []) {
...
-            return sprintf(_n('%d second', '%d seconds', $value), $value);
+            return sprintf(_n('%s second', '%s seconds', $value), $formatted_number);
 
          case 'millisecond' :
             //TRANS: %d is a number of milliseconds
-            return sprintf(_n('%d millisecond', '%d milliseconds', $value), $value);
+            return sprintf(_n('%s millisecond', '%s milliseconds', $value), $formatted_number);
 
          case 'auto':
                $value = str_replace([' ', '&nbsp;'], ['', ''], $value); // unformat value
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1556]: I guess this could be removed now.

File: inc/dropdown.class.php [hunk 2]
```diff
@@ -1513,41 +1513,44 @@ static function showNumber($myname, $options = []) {
     *
     * @param $value   integer   number of item
     * @param $unit    string    of unit (maybe year, month, day, hour, % for standard management)
+    * @param $decimals integer  number of decimal
    **/
-   static function getValueWithUnit($value, $unit) {
+   static function getValueWithUnit($value, $unit, $decimals = null) {
 
       if (strlen($unit) == 0) {
          return $value;
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1521]: Here you should now return the formatted number.

**Review by cedric-anne — COMMENTED** [review]
1. You should also use the formatted number for `'%'` and `default` cases in `Dropdown::getValueWithUnit()`.

2. You should remove the call to `Html::formatNumber()` in `decimal` case in `Search::giveItem()`.

3. You sould remove calls to `Html::formatNumber()` for `count`, `number` and `decimal` in `CommonDBTM::getValueToDisplay()`.

File: inc/dropdown.class.php (hunk 1 above)
Comment (stonebuzz) [reply inc/dropdown.class.php:1556]: ```millisecond``` case ?

**Review by orthagh — APPROVED** [review]

File: inc/commondbtm.class.php [hunk 3]
```diff
@@ -4556,13 +4556,13 @@ function getValueToDisplay($field_id_or_search_options, $values, $options = [])
...
                   if ($options['html']) {
-                     return Dropdown::getValueWithUnit(Html::formatNumber($value, false, 0), $unit);
+                     return Dropdown::getValueWithUnit($value, $unit);
                   }
                   return $value;
 
                case "decimal" :
                   if ($options['html']) {
-                     return Dropdown::getValueWithUnit(Html::formatNumber($value), $unit);
+                     return Dropdown::getValueWithUnit($value, $unit, true);
```
Comment (cedric-anne) [inline inc/commondbtm.class.php:4565]:
```suggestion
                     return Dropdown::getValueWithUnit($value, $unit);
```

File: inc/commondbtm.class.php [hunk 4]
```diff
@@ -4556,13 +4556,13 @@ function getValueToDisplay($field_id_or_search_options, $values, $options = [])
                      return $searchoptions['toadd'][$value];
                   }
                   if ($options['html']) {
-                     return Dropdown::getValueWithUnit(Html::formatNumber($value, false, 0), $unit);
+                     return Dropdown::getValueWithUnit($value, $unit);
```
Comment (cedric-anne) [inline inc/commondbtm.class.php:4559]:
```suggestion
                     return Dropdown::getValueWithUnit($value, $unit, 0);
```

File: inc/dropdown.class.php [hunk 5]
```diff
@@ -1513,48 +1513,51 @@ static function showNumber($myname, $options = []) {
...
       }
 
+      $formatted_number = Html::formatNumber($value, false, $decimals !== null ? $decimals : -1);
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1524]:
```suggestion
      $formatted_number = Html::formatNumber($value, false, $decimals !== null ? $decimals : -1);

      if (strlen($unit) == 0) {
         return $formatted_number;
      }
```

File: inc/dropdown.class.php [hunk 6]
```diff
@@ -1513,48 +1513,51 @@ static function showNumber($myname, $options = []) {
...
+            //TRANS: %s is a number of milliseconds
+            return sprintf(_n('%s millisecond', '%s milliseconds', $value), $formatted_number);
 
          case 'auto':
                $value = str_replace([' ', '&nbsp;'], ['', ''], $value); // unformat value
                return Toolbox::getSize($value*1024*1024);
 
          case '%' :
-            return sprintf(__('%d%%'), $value);
+            return sprintf(__('%s'), $value);
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1560]:
```suggestion
            return sprintf(__('%s'), $formatted_number);
```

File: inc/dropdown.class.php [hunk 7]
```diff
@@ -1513,48 +1513,51 @@ static function showNumber($myname, $options = []) {
...
 
          default :
             return sprintf(__('%1$s %2$s'), $value, $unit);
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1563]:
```suggestion
            return sprintf(__('%1$s %2$s'), $formatted_number, $unit);
```

File: inc/dropdown.class.php [hunk 8]
```diff
@@ -1513,48 +1513,51 @@ static function showNumber($myname, $options = []) {
...
+            return sprintf(_n('%s second', '%s seconds', $value), $formatted_number);
 
          case 'millisecond' :
-            //TRANS: %d is a number of milliseconds
-            return sprintf(_n('%d millisecond', '%d milliseconds', $value), $value);
+            //TRANS: %s is a number of milliseconds
+            return sprintf(_n('%s millisecond', '%s milliseconds', $value), $formatted_number);
 
          case 'auto':
                $value = str_replace([' ', '&nbsp;'], ['', ''], $value); // unformat value
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1556]:
```suggestion
```

File: inc/search.class.php [hunk 9]
```diff
@@ -6434,9 +6432,7 @@ static function giveItem($itemtype, $ID, array $data, $meta = 0,
                            && isset($so['toadd'][$data[$ID][$k]['name']])) {
                         $out .= $so['toadd'][$data[$ID][$k]['name']];
                      } else {
-                        $number = str_replace(' ', '&nbsp;',
-                                              Html::formatNumber($data[$ID][$k]['name']));
-                        $out   .= Dropdown::getValueWithUnit($number, $unit);
+                        $out .= Dropdown::getValueWithUnit($data[$ID][$k]['name'], $unit, true);
```
Comment (cedric-anne) [inline inc/search.class.php:6435]:
```suggestion
                        $out .= Dropdown::getValueWithUnit($data[$ID][$k]['name'], $unit);
```

File: inc/commondbtm.class.php (hunk 4 above)
Comment (stonebuzz) [reply inc/commondbtm.class.php]: done

File: inc/dropdown.class.php (hunk 5 above)
Comment (stonebuzz) [reply inc/dropdown.class.php]: done

File: inc/dropdown.class.php (hunk 7 above)
Comment (stonebuzz) [reply inc/dropdown.class.php]: done

File: inc/dropdown.class.php (hunk 8 above)
Comment (stonebuzz) [reply inc/dropdown.class.php]: done

File: inc/dropdown.class.php (hunk 6 above)
Comment (stonebuzz) [reply inc/dropdown.class.php]: done

File: inc/search.class.php (hunk 9 above)
Comment (stonebuzz) [reply inc/search.class.php]: done

File: inc/dropdown.class.php [hunk 10]
```diff
@@ -1513,51 +1513,53 @@ static function showNumber($myname, $options = []) {
...
+            return sprintf(_n('%s millisecond', '%s milliseconds', $value), $formatted_number);
 
          case 'auto':
-               $value = str_replace([' ', '&nbsp;'], ['', ''], $value); // unformat value
-               return Toolbox::getSize($value*1024*1024);
+            return Toolbox::getSize($value*1024*1024);
 
          case '%' :
-            return sprintf(__('%d%%'), $value);
+            return sprintf(__('%s'), $formatted_number);
```
Comment (cedric-anne) [inline inc/dropdown.class.php:1559]:
```suggestion
            return sprintf(__('%s%%'), $formatted_number);
```

Comment (stonebuzz) [reply inc/dropdown.class.php]: done

**Review by cedric-anne — APPROVED** [review]

**Review by trasher — DISMISSED** [review]

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**